pyth/client/src/client_cache.rs

```diff
@@ -0,0 +1,189 @@
+use {
+    crate::{client::PythClientTrait, error, PythClient},
+    async_stream::stream,
+    async_trait::async_trait,
+    grug::{Binary, Inner, Lengthy, NonEmpty},
+    indexer_disk_saver::{error::Error, persistence::DiskPersistence},
+    reqwest::{IntoUrl, Url},
+    std::{
+        collections::HashMap,
+        env,
+        path::{Path, PathBuf},
+        sync::{
+            atomic::{AtomicU64, Ordering},
+            Arc,
+        },
+        thread::sleep,
+        time::Duration,
+    },
+    tokio::runtime::Runtime,
+    tokio_stream::StreamExt,
+};
+
+#[derive(Debug, Clone)]
+pub struct PythClientCache {
+    base_url: Url,
+    // Used to return newer vaas at each call.
+    vaas_index: Arc<AtomicU64>,
+}
+
+impl PythClientCache {
+    pub fn new<U: IntoUrl>(base_url: U) -> Result<Self, error::Error> {
+        Ok(Self {
+            base_url: base_url.into_url()?,
+            vaas_index: Arc::new(AtomicU64::new(0)),
+        })
+    }
+
+    /// Load data from cache or retrieve it from the source.
+    pub fn load_or_retrieve_data<I>(
+        base_url: Url,
```

Comment:
Since we're cloning `base_url` inside this method (because of the loop) you can change that to `&Url` and avoid base_url clones in the methods calling `load_or_retrieve_data`